### PR TITLE
Fix emitting of function reference types without GC

### PR DESCRIPTION
**Pull request by tlively**
We previously had logic to emit GC types used in the IR as their corresponding
top types when GC was not enabled (so e.g. nullfuncref would be emitted as
funcref), but the logic was not robust enough and non-null function references
were not properly emitted as funcref.

Refactor the relevant code to be more robust and future-proof, and add a test
demonstrating that the lowering works as intended.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5737** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5737" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5737?utm_source=stack-comment).

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -1393,6 +1393,18 @@ void WasmBinaryWriter::writeInlineBuffer(const char* data, size_t size) {
 
 void WasmBinaryWriter::writeType(Type type) {
   if (type.isRef()) {
+    // The only reference types allowed without GC are funcref and externref. We
+    // internally use more refined versions of those types, but we cannot emit
+    // thoes more refined types.
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp:1398]:
```suggestion
    // those more refined types.
```

File: test/lit/downgrade-reftypes.wast [hunk 2]
```diff
@@ -0,0 +1,61 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; Write to a binary, lowering away refined GC types.
+;; RUN: wasm-as %s -all --disable-gc -g -o %s.wasm
+
+;; Read it back and verify that the types were lowered away.
+;; RUN: wasm-dis %s.wasm -all -o - | filecheck %s
```
Comment (kripken) [inline test/lit/downgrade-reftypes.wast:7]: You can just do `wasm-opt --roundtrip` for this kind of thing.

**Review by kripken — APPROVED** [review]
lgtm % (comments + fuzzing)

File: test/lit/downgrade-reftypes.wast (hunk 2 above)
Comment (kripken) [reply test/lit/downgrade-reftypes.wast:7]: Though maybe as it is is more explicit? I don't feel strongly.

Comment (tlively) [reply test/lit/downgrade-reftypes.wast:7]: In particular I want to write without GC enabled and then read with GC enabled. I don't think --roundtrip would let me do that.

Comment (kripken) [reply test/lit/downgrade-reftypes.wast:7]: Ah, good point...

**Pull request merged**